### PR TITLE
Improves threshold_signature test

### LINKED ISSUE #6: Pending test related suggestions from the PoC PR
Following are the pending test related suggestions from the PoC PR https://github.com/maidsafe/BLS-DKG/pull/2 that shall be addressed asap:

1, Comment on tests.rs lines 213 to 235
Ideal for making this test more robust: iterate over all THRESHOLD + 1 - sized subsets of sig_shares 
and check that they produce valid signatures and also that the signatures 

2, https://github.com/maidsafe/BLS-DKG/pull/2#discussion_r407961553
We should also test that combining less that THRESHOLD + 1 shares is not enough.

3, https://github.com/maidsafe/BLS-DKG/pull/2#discussion_r407963213
Same here: we should test that this work with any THRESHOLD + 1 sized subset of dec_shares and that it doesn't work if we have less that THRESHOLD + 1 shares.

### PR DESCRIPTION
**Pull request by Yoga07**
Resolves maidsafe/BLS-DKG#6

### CODE REVIEW

**Review by maqi — DISMISSED** [review]

**Review by madadam — DISMISSED** [review]
Looks good. We should give the same treatment to the threshold encryption test. And also have tests for the negative cases: that less than threshold + 1 shares is not enough. But I suppose those are coming in a future PR?

**Comment by Yoga07** [comment]
Thanks for the reviews @madadam @maqi. I've updated more tests to use the same combinations pattern as suggested in the review comments. Would be nice to get one last :heavy_check_mark: :)

**Review by maqi — DISMISSED** [review]

File: src/key_gen/tests.rs [hunk 1]
```diff
@@ -256,29 +257,41 @@ fn threshold_signature() {
...
-        .combine_signatures(
-            sig_shares
-                .iter()
-                .cycle()
-                .skip(sig2_start_idx)
-                .take(THRESHOLD + 1),
-        )
-        .expect("signature shares match");
-    assert!(pub_key_set.public_key().verify(&sig2, msg));
+    let deficient_sig_combinations = sig_shares.iter().combinations(THRESHOLD - 1);
```
Comment (madadam) [inline src/key_gen/tests.rs]: Wouldn't it make more sense to test with `THRESHOLD` instead of `THRESHOLD - 1` because it's the maximum number of shares that is still not enough?

File: src/key_gen/tests.rs [hunk 2]
```diff
@@ -323,6 +336,23 @@ fn threshold_encrypt() {
         })
         .collect();
 
-    let decrypted = pub_key_set.decrypt(&dec_shares, &ciphertext).unwrap();
-    assert_eq!(msg, decrypted.as_slice());
+    // Test threshold encryption verification for a combination of shares - should pass as there
+    // are THRESHOLD + 1 shares aggregated in each combination
+    let dec_share_combinations = dec_shares.iter().combinations(THRESHOLD + 1);
```
Comment (madadam) [inline src/key_gen/tests.rs:341]: Same here: better use `THRESHOLD`, not `THRESHOLD - 1`.

File: src/key_gen/tests.rs [hunk 3]
```diff
@@ -323,6 +336,23 @@ fn threshold_encrypt() {
...
+    for dec_share in dec_share_combinations {
+        let decrypted = pub_key_set.decrypt(dec_share, &ciphertext).unwrap();
+        assert_eq!(msg, decrypted.as_slice());
+    }
+
+    // Test threshold decryption for a combination of shares - shouldn't decrypt as there
+    // are THRESHOLD - 1 shares aggregated in each combination
+    let deficient_dec_share_combinations = dec_shares.iter().combinations(THRESHOLD - 1);
```
Comment (madadam) [inline src/key_gen/tests.rs]: Same here: better use `THRESHOLD`, not `THRESHOLD - 1`.

**Review by madadam — APPROVED** [review]
